File: app/schemas/playlist.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field
from app.schemas.song import SongResponse


class PlaylistCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1024)
    is_public: bool = False
    is_collaborative: bool = False
    cover_url: Optional[str] = None


class PlaylistUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=1024)
    is_public: Optional[bool] = None
    is_collaborative: Optional[bool] = None
    cover_url: Optional[str] = None


class PlaylistSongItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    playlist_id: str
    song_id: str
    position: int
    added_at: datetime
    song: Optional[SongResponse] = None


class PlaylistResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str
    title: str
    description: Optional[str] = None
    is_public: bool
    is_collaborative: bool
    cover_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    song_count: int = 0
    songs: Optional[List[PlaylistSongItem]] = None


class AddPlaylistSongRequest(BaseModel):
    song_id: str
    position: Optional[int] = None


class ReorderPlaylistSongsRequest(BaseModel):
    song_ids: List[str] = Field(..., description="Ordered list of song IDs representing the new playlist sequence")

File: mobile_app/lib/shared/models/playlist.dart
import 'song.dart';

class PlaylistSongItem {
  final String id;
  final String playlistId;
  final String songId;
  final int position;
  final Song? song;

  const PlaylistSongItem({
    required this.id,
    required this.playlistId,
    required this.songId,
    required this.position,
    this.song,
  });

  factory PlaylistSongItem.fromJson(Map<String, dynamic> json) {
    return PlaylistSongItem(
      id: json['id']?.toString() ?? '',
      playlistId: json['playlist_id']?.toString() ?? '',
      songId: json['song_id']?.toString() ?? '',
      position: (json['position'] as num?)?.toInt() ?? 0,
      song: json['song'] is Map
          ? Song.fromJson((json['song'] as Map).cast<String, dynamic>())
          : null,
    );
  }
}

class Playlist {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final bool isPublic;
  final bool isCollaborative;
  final String? coverUrl;
  final int songCount;
  final List<PlaylistSongItem> songs;

  const Playlist({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    this.isPublic = false,
    this.isCollaborative = false,
    this.coverUrl,
    this.songCount = 0,
    this.songs = const [],
  });

  factory Playlist.fromJson(Map<String, dynamic> json) {
    return Playlist(
      id: json['id']?.toString() ?? '',
      userId: json['user_id']?.toString() ?? '',
      title: json['title']?.toString() ?? 'Untitled playlist',
      description: json['description']?.toString(),
      isPublic: json['is_public'] == true,
      isCollaborative: json['is_collaborative'] == true,
      coverUrl: json['cover_url']?.toString(),
      songCount: (json['song_count'] as num?)?.toInt() ?? 0,
      songs: (json['songs'] as List?)
              ?.map((e) => PlaylistSongItem.fromJson((e as Map).cast<String, dynamic>()))
              .toList() ??
          const [],
    );
  }
}
